src/peft/tuners/oft/layer.py

```diff
@@ -282,100 +425,323 @@ def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
+                oft_s = self.oft_s[active_adapter]
+                dropout = self.oft_dropout[active_adapter]
 
-                module_dropout = self.module_dropout[active_adapter]
-
-                # Modify current execution weights
-                if (not self.training) or (self.training and torch.rand(1) > module_dropout):
-                    result = self._get_delta_activations(active_adapter, result, *args, **kwargs)
+                rank = self.r[active_adapter]
+                coft = self.coft[active_adapter]
+                eps = self.eps[active_adapter]
 
-            if base_bias is not None:
-                result = result + base_bias.data
-            if len(result.shape) == 4:
-                result = result.permute(0, 3, 1, 2)
+                if coft:
+                    with torch.no_grad():
+                        oft_r.copy_(self._project_batch(oft_r, eps=eps))
 
-        result = result.to(previous_dtype)
-        return result
+                orth_rotate = self._cayley_batch(oft_r)
+                orth_rotate = dropout(orth_rotate)
+                oft_mat = self._block_diagonal(orth_rotate, rank)
 
+                oft_rotation = oft_mat @ oft_rotation
+                oft_scale = oft_s * oft_scale
 
-class Linear(OFTLayer):
-    """OFT implemented in Linear layer"""
-
-    def __init__(
-        self,
-        base_layer: nn.Module,
-        adapter_name: str = "default",
-        r: int = 0,
-        module_dropout: float = 0.0,
-        init_weights: bool = True,
-        **kwargs,
-    ):
-        super().__init__(base_layer)
+            x = x.to(self.get_base_layer().weight.data.dtype)
 
-        # Create adapter and set it active
-        self._active_adapter = adapter_name
-        self.update_layer(adapter_name, r, module_dropout, init_weights, **kwargs)
+            orig_weight = self.get_base_layer().weight.data
+            orig_weight = torch.transpose(orig_weight, 0, 1)
+            oft_rotation = oft_rotation.to(previous_dtype)
+            orig_weight = orig_weight.to(previous_dtype)
+            rotated_weight = torch.mm(oft_rotation, orig_weight)
+            rotated_weight = torch.transpose(rotated_weight, 0, 1)
 
-    def _get_delta_activations(
-        self, adapter_name: str, input: torch.Tensor, *args: Any, **kwargs: Any
-    ) -> torch.Tensor:
-        delta_weight = self.get_delta_weight(adapter_name)
+            scaled_rotated_weight = rotated_weight * oft_scale
 
-        base_layer = self.get_base_layer()
-        base_weight = base_layer.weight.data
-        delta_weight = delta_weight[: base_weight.shape[0], : base_weight.shape[0]]
+            scaled_rotated_weight = scaled_rotated_weight.to(previous_dtype)
+            if self.base_layer.bias is not None:
+                self.base_layer.bias = self.base_layer.bias.to(previous_dtype)
+            result = F.linear(input=x, weight=scaled_rotated_weight, bias=self.base_layer.bias)
```

Comment:
Let's not mutate the bias of the base layer.

```suggestion
            bias = self.get_base_layer().bias.to(previous_dtype) if self.get_base_layer().bias is not None else None
            result = F.linear(input=x, weight=scaled_rotated_weight, bias=bias)
```